### ISSUE TITLE
Check for Stack Smashing Protection support in the target libc

### ISSUE THREAD

**Issue by lb90** [mesonbuild/meson#10673]
Meson should check if the target libc provides stack smashing protection routines. If not, it should automatically add a dependency on the generic stack smashing protection routines in libssp.

See https://gitlab.gnome.org/GNOME/gtksourceview/-/issues/285#note_1524407

**Comment by lb90** [comment]
@xclaesse Do you know if it's possible to add a lib _after_ the default libs that gcc adds? If that's possible, we could simply add -lssp at the very end?

**Comment by xclaesse** [comment]
Not sure to understand the problem, it does not work simply `cc.find_library('ssp', required: false)` and add that to your target dependencies ?

**Comment by lb90** [comment]
He @xclaesse!

`libssp` provides generic implementations, ideally one should use the implementations provided by the target libc, if there are any. If one adds the ssp dependency blindly, I assume that the implementations from libssp will always be used, rather than the libc's?

**Comment by xclaesse** [comment]
Something like that?
```
ssp_dep = []
if not cc.has_function('__stack_chk_fail') # or whatever compiler checks can tell if ssp works without libs
  ssp_dep = cc.find_library('ssp', required: false)
endif
```

Sorry if I'm too naive, I know nothing about the situation :/

**Comment by lb90** [comment]
Thanks. That might work, indeed, I'll make some tests. Ideally that should be done automatically by Meson though

**Comment by xclaesse** [comment]
We could add custom support for `dependency('ssp')` like we have for a few other special dependencies (e.g. iconv).

**Comment by eli-schwartz** [comment]
I find this topic inherently troubling.

The implication here is that the compiler itself is inherently broken in cases that include compiling a special source file with no options, but only with the mingw compiler. And it does it for, effectively, non-public implementation details of the runtime.

It feels like the proper fix here is that mingw needs to do the right thing in its own internal spec file.

**Comment by lb90** [comment]
Hi, @eli-schwartz!

If I understand it correctly, you mean that GCC and CLang should add `libssp` as part of the standard libs added automatically, right?

In theory it's not just mingw-w64, even other libc's may not provide stack protection routines, because those are not part of any standard (C or POSIX). See for example https://embeddedartistry.com/blog/2020/05/18/implementing-stack-smashing-protection-for-microcontrollers-and-embedded-artistrys-libc/

**Comment by eli-schwartz** [comment]
No, what I'm saying is that that is ***already*** the case for gcc/clang, everywhere but mingw.

And the embeddedartistry post does NOT count... because they specifically talk about the use of -nostdlib.

To me, it's quite simple. The libc vendor knows what is and isn't needed. A compiler, when used with its default libc, should know what is and isn't needed. When using one compiler flag that activates stack smashing protection, you shouldn't need to pass other flags as ***bookkeeping*** just so it works. No more than gcc requires you to use -lgcc, -lgcc_s, or -lc, under the various circumstances that it's internally handled.


Alternative libcs aren't unusually special, you can write a custom spec file for them to do various magic or you can hardcode it into your compile commands. Musl, for example, has its own spec file (which links with -nostdlib and inserts its own implementation instead). Embedded libcs may often prefer to hardcode the compile commands, fair enough.

But mingw is apparently just outright broken, because without using -nostdlib the default link is broken.

Broken!

**Comment by lb90** [comment]
Ah, interesting! Looks like spec files are perfect for the task, indeed: https://gcc.gnu.org/onlinedocs/gcc/Spec-Files.html

I'm going to prepare a patch for mingw GCC, then!

**Comment by mstorsjo** [comment]
FWIW, when passing `-fstack-protector-strong` to the compiler driver when linking, it already does the right thing, for both GCC and clang.

But since some libcs provide the same functions, you can get away with not using that flag on many platforms.

**Comment by mstorsjo** [comment]
IMO, the correct way forward would be for mingw-w64 to provide the functions that the fortify macros invoke, themselves in the mingw libraries that always are linked in. Technically, I guess these don’t even need to match the function names used by the compiler’s libssp, they could be entirely separate ones. I’ve meant to implement this myself some time, but haven’t quite gotten to it.

**Comment by lb90** [comment]
Thank you you very much, @mstorsjo!

No rush, of course! 😉 take all the time you need! Would it be possible to merge libssp in mingw-w64 libraries as it currently is, as a start?

**Comment by mstorsjo** [comment]
> No rush, of course! 😉 take all the time you need! Would it be possible to merge libssp in mingw-w64 libraries as it currently is, to start?

Not as a quick action just to get started, no - libssp as is today is GCC source, which is licensed slightly differently than the rest of mingw-w64 etc.

But can you confirm what you want to have fixed here - for actual use of the compiler generated stack smashing protection, i.e. compiling with `-fstack-protector-strong`, the fact that this requires you to consistently pass the same flag to both compile and link commands - is that bit acceptable as is? Sure, it would be more convenient if libssp was integrated in the default libraries like on many Unix libcs though.

Then for the use of `_FORTIFY_SOURCE`, which currently requires you to manually link in libssp - yes, that really should be fixed. From a quick look at it right now, it shouldn't even be a big thing to do...

**Comment by lb90** [comment]
> Not as a quick action just to get started, no - libssp as is today is GCC source, which is licensed slightly differently than the rest of mingw-w64 etc.

Got it, thanks!

> compiling with -fstack-protector-strong, the fact that this requires you to consistently pass the same flag to both compile and link commands - is that bit acceptable as is?

Nice! Yes, that's quite ok!

> Then for the use of _FORTIFY_SOURCE, which currently requires you to manually link in libssp - yes, that really should be fixed. From a quick look at it right now, it shouldn't even be a big thing to do...

Yeah, we can wait until there's proper support in mingw-w64, no problem at all! :)

**Comment by mstorsjo** [comment]
> IMO, the correct way forward would be for mingw-w64 to provide the functions that the fortify macros invoke, themselves in the mingw libraries that always are linked in. Technically, I guess these don’t even need to match the function names used by the compiler’s libssp, they could be entirely separate ones.

Regarding this, I had kinda misrememberd the details; libssp consists of 17 exported symbols - 2 of them (`__stack_chk_guard` and `__stack_chk_fail`) are relevant for the compiler generated instrumentation for stack smashing protection (`-fstack-protector-strong`), while the other 15 are functions that are called by the system headers from fortification macros. Out of those, only 9 functions are currently ever called by the mingw header fortification macros.

In any case, I finally took the plunge and made reimplementations of all of them - see https://sourceforge.net/p/mingw-w64/mailman/message/37707134/ and https://sourceforge.net/p/mingw-w64/mailman/message/37707132/.

**Comment by lb90** [comment]
Thank you all, @xclaesse @eli-schwartz @mstorsjo. Outstanding work!!!

I'm going to close this issue then :)
